### PR TITLE
Remove code duplication plus other fixes

### PR DESCRIPTION
**Pull request by VSEphpbb**

### CODE REVIEW

File: controller/base.php [hunk 1]
```diff
@@ -83,4 +83,29 @@ public function is_available()
...
 		return (bool) $this->config['ideas_forum_id'] && (bool) $this->config['ideas_poster_id'];
 	}
+
+	/**
+	 * Assign idea lists template variables
+	 *
+	 * @param string $block The template block var name
+	 * @param array  $rows  The Idea row data
+	 */
+	protected function assign_template_vars($block, $rows)
```
Comment (rxu) [inline controller/base.php]: I'd probably rename it to the something like ```assign_template_block_vars``` as it only assigns template blocks.

**Comment by VSEphpbb** [comment]
@michaelcullum :smile:

**Comment by VSEphpbb** [comment]
Updated with fixes for all the Status grouping/sorting issues.

File: factory/ideas.php [hunk 2]
```diff
@@ -205,36 +230,31 @@ public function get_ideas($number = 10, $sort = 'date', $sort_direction = 'DESC'
...
+				),
+				array(
+					'FROM'	=> array($this->table_rfcs => 'r'),
+					'ON'	=> 'i.idea_id = r.idea_id',
+				),
+			),
+			'WHERE'			=> 'i.idea_id = ' . (int) $id,
+		);
+
+		$sql = $this->db->sql_build_query('SELECT', $sql_array);
```
Comment (rxu) [inline factory/ideas.php]: If the idea data is stable enough, we could cache the query result.

File: factory/ideas.php [hunk 3]
```diff
@@ -205,36 +230,31 @@ public function get_ideas($number = 10, $sort = 'date', $sort_direction = 'DESC'
 	 */
 	public function get_idea($id)
```
Comment (rxu) [inline factory/ideas.php]: Probably ```get_idea_data``` would be more accurate name?

File: factory/ideas.php [hunk 4]
```diff
@@ -247,8 +267,8 @@ public function get_idea($id)
 	public function get_status_from_id($id)
 	{
 		$sql = 'SELECT status_name
-			FROM ' . $this->table_statuses . "
-			WHERE status_id = $id";
+			FROM ' . $this->table_statuses . '
+			WHERE status_id = ' . (int) $id;
 		$result = $this->db->sql_query_limit($sql, 1);
```
Comment (rxu) [inline factory/ideas.php]: Maybe cache the query result as well?

**Comment by rxu** [comment]
Impressive :+1:

File: factory/ideas.php (hunk 3 above)
Comment (VSEphpbb) [inline factory/ideas.php]: I prefer get_idea, as it was

File: factory/ideas.php (hunk 2 above)
Comment (VSEphpbb) [inline factory/ideas.php]: But there are a lot of operations that update the idea... Voting and mod operations.

Comment (rxu) [inline factory/ideas.php]: I've found 6 pieces of ```'UPDATE ' . $this->table_ideas``` so far, all in ideas factory:
```change_status```, ```set_title```, ```vote``` (twice), ```remove_vote``` and ```submit```.

Comment (VSEphpbb) [inline factory/ideas.php]: But it is also an issue when deleting an idea, for example. There are a number of places where this query is called upon and I just want to be sure it is always fresh without having to spread cache cleanups all around ;)

**Comment by rxu** [comment]
I think this should be merged to let further fixing (a lot of things more).

**Comment by VSEphpbb** [comment]
@michaelcullum bumpies

**Comment by VSEphpbb** [comment]
@michaelcullum can we get this PR looked at when you have time. It's holding up further progress on this extension.

**Comment by michaelcullum** [comment]
Sorry yeah. I'll look over this tonight when I get home (Remind me)

**Comment by VSEphpbb** [comment]
I always remind you :smile:

File: controller/index_controller.php [hunk 5]
```diff
@@ -29,53 +30,17 @@ public function index()
...
-				'AUTHOR'	=> $this->link_helper->get_user_link($row['idea_author']),
-				'DATE'		=> $this->user->format_date($row['idea_date']),
-				'READ'      => $row['read'],
-				'VOTES_UP'	=> $row['idea_votes_up'],
-				'VOTES_DOWN'=> $row['idea_votes_down'],
-				'POINTS'    => $row['idea_votes_up'] - $row['idea_votes_down'],
-			));
-		}
+		// Generate latest ideas
+		$ideas = $this->ideas->get_ideas(self::IDEAS, 'date', 'DESC');
```
Comment (michaelcullum) [inline controller/index_controller.php]: Can we rename this constant btw so we know what it is inline?

**Pull request merged**